### ISSUE TITLE
Proposal: Add GPU Details to TensorFlow data store

### ISSUE THREAD

**Issue by cvinson830** [tensorflow/tensorflow#20597]
We’re looking to extend the TensorFlow and TensorBoard user interface to display interesting GPU details for each node in a neural network. The GPU Details that we plan to display focus on the usage of tensor cores. Tensor cores enable deep learning applications to run significantly faster.

The UI TensorBoard proposal is here: https://github.com/tensorflow/tensorboard/issues/1271

In order to display the interesting GPU Details information in TensorBoard, we need to add the GPU Details to a data store.  One idea is for NodeDef to encapsulate a new gpu_details protobuf.  See proposed protobuf definitions below.

 Update existing NodeDef
```
message NodeDef {
  string name = 1;
  string op = 2;
  repeated string input = 3;
  string device = 4;
  map<string, AttrValue> attr = 5;

  GpuDetails gpu_details = 6;		// new, optional
};
```

Create new GpuDetails protobuf
```
message GpuDetails {
  repeated GpuKernelAttrs gpuKernelAttrs = 1;
};
```

Create new GpuKernelAttrs protobuf
```
message GpuKernelAttrs {
  string kernel_name = 1;
  int32 calls = 2;
  float total_duration = 3;
  float average_duration = 4;
  float minimum_duration = 5;
  float maximum_duration = 6; 
}
```

We're aware that this is a change to a fundamental data structure.  Another potential solution we considered is adding a new event type in TF that captures the GPU Details information.

We're open to other suggestions you have - any guidance here would be appreciated.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by cvinson830** [comment]
Hello, here are the responses to prompts:
```
Have I written custom code: No
OS Platform and Distribution: Windows, Linux, MAC
TensorFlow installed from: github
TensorFlow version: 1.9
Bazel version: bazel release 0.13.0
CUDA/cuDNN version: V9.0.176/7.0.4
GPU model and memory: TITAN V 
Exact command to reproduce: N/A
```

**Comment by asimshankar** [comment]
Thanks for reaching out @cvinson830 

At first glance, it seems that these stats are more suitable for [`NodeExecStats`](https://github.com/tensorflow/tensorflow/blob/4297b9d492327072c0d64b7243925eba192fe028/tensorflow/core/framework/step_stats.proto#L53) rather than `NodeDef`, which is what [TensorBoard uses to render memory stats](https://github.com/tensorflow/tensorboard/blob/3f9c511ef1b6fb45a3fd11ecddde7442ce1771f9/tensorboard/plugins/graph/tf_graph_common/graph.ts#L436)

We do already collect some profiling information for GPUs, see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/client/timeline.py - which generates a chrome trace formatted timeline from the `StepStats` protocol buffer.

I'd have to think more about the details here, but hopefully these pointers are useful.
Could you also add some detail on the kind of information we'd collect and how?
CC @zheng-xq - who has more knowledge and experience around GPU profiling and stats than I do.

**Comment by jart** [comment]
This sounds good to me. `NodeExecStats` is probably the right place for this data, and it shouldn't cost us anything to add additional fields.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by cvinson830** [comment]
Hi, yes this is still in the works.  This sprint I'm going to work on putting this information in the NodeExecStats protobuf.  I'll likely submit the pull request to TF and TB this month.

**Comment by cvinson830** [comment]
@asimshankar Hi Asim, we'd like to augment a node in a neural network with this extra gpu information.  The node is associated with the NodeDef protobuf.  Question: How is a node in a neural network associated with NodeExecStats?

**Comment by cvinson830** [comment]
@asimshankar 
+ @jart 

A little more detail may help: 

The GraphDef protobuf (graph.proto) is the topmost object in our TB event file.  When we create a TB event file and load it in TB, we see the GRAPHS tab selected, and our neural network rendered properly.

GraphDef contains a repeated list of NodeDef protobufs (node_def.proto).  This was the basis for the proposal of adding gpu_details protobuf to NodeDef.

Your guidance is to not use NodeDef protobuf, and use NodeExecStats instead.  I'm happy to take your advice.  However, I do not see where the NodeExecStats protobuf (step_stats.proto) gets added to the graphdef protobuf.  

Question: What is the relationship between NodeDef and NodeExecStats?  In other words, how can I fill a gpu_details protobuf inside NodeExecStats object and relate it to each NodeDef in a GraphDef?

**Comment by dzier** [comment]
@asimshankar or @jart , would either of you be able to answer @cvinson830's questions above. Currently, this issue is blocking us from using the suggested NodeExecStats.

**Comment by asimshankar** [comment]
Sorry, somehow missed @cvinson830 's update.

The GraphDef protobuf describes the computation, it does not contain any runtime statistics.
NodeExecStats is filled in when tracing is turned on in a call to `Session::Run` and includes a bunch of information about the execution of the node. NodeExecStats for each node of the graph that was executed in the `Session::Run` call is included in the StepStats proto message and the node is linked to the corresponding node in the graph by the string name of the node.

https://github.com/tensorflow/tensorflow/blob/e079a8ab1eea2507018770f1e9fe5d5d2793c9c7/tensorflow/core/framework/step_stats.proto#L58

For example:

```python
import tensorflow as tf

x = tf.random_uniform([2, 2])
y = tf.reduce_sum(x)

with tf.Session() as sess:
  metadata = tf.RunMetadata()
  sess.run(y, options=tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE), run_metadata=metadata)
  print(metadata) # This will include NodeExecStats for each node on each device
```

As mentioned earlier, the node name is `NodeExecStats` is used to associate the stats with a node in the graph and render information on Tensorboard (see https://github.com/tensorflow/tensorboard/blob/3f9c511ef1b6fb45a3fd11ecddde7442ce1771f9/tensorboard/plugins/graph/tf_graph_common/graph.ts#L436)

Let me know if that helps. Thanks.

**Comment by cvinson830** [comment]
@asimshankar 

In our workflow, the network file is not created via TensorFlow's  Session::Run call via python.  Rather, it is created using the protobuf methods in the GraphDef protobuf in the file "graph.proto" via C++.  The kernel information (ie, the gpu_details protobuf) is obtained from an external data source, and is added to the graphdef object outside of the TensorFlow framework.

Given that flow, is there an alternate way to connect StepStats to the graphdef protobuf?  Or is it ok to encapsulate the gpu_details within the node_def protobuf?  (I have the latter scenario working, fwiw).

**Comment by asimshankar** [comment]
I may be misunderstanding something - isn't the information being added related to _execution_ of the graph? So, which `GraphDef` is being modified in the prototype you have? For example, let's say I'm a Python user that does something like this:

```python
import tensorflow as tf

y = tf.matmul([[1.]], [[2.]])
with tf.Session() as sess:
  sess.run(y)
```
How will I get access to the `GpuDetails` protobuf?

(I'm probably missing something, but I still don't quite understand why these stats are a property of the `GraphDef`, which describes the computation to execute)

**Comment by cvinson830** [comment]
@asimshankar

From 30,000' our program does these three things:
1) Read in existing TensorFlow network file 
2) Correlate our gpuDetails business data with each node in the network
3) Write out TensorBoard event file.

The user does not get access to or manipulate GpuDetails protobuf.
Instead, our program is injecting our gpuDetails data into each node on an existing graph.

**Comment by asimshankar** [comment]
In that case, I wonder if it would be cleaner to separate the two by:

1. Enhancing the `GpuKernelAttrs` message with a `node_name` that will link the stats with a precise node in the graph.

2. Write it out as a separate, self-contained file.

3. Write a TensorBoard plugin that can consume this file (and use the `node_name` to correlate with the `GraphDef`).

This way the plugin and the tooling is self contained. No need to change `StepStats` or `GraphDef`
Does that sound reasonable? (If not, perhaps an in-person meeting/video-conference would be more effective than going back and forth over email/comments. Though, happy to chat either way)

**Comment by dzier** [comment]
I agree with the separation. It does make a cleaner flow. However, there is a stipulation that we may not have an event file at all then. Our product is creating performing the correlation and generating the complete event file that is then consumed by Tensorboard. 

However, I agree that it would be more productive if we set up a meeting/video-conference. Is there a time that works well for you this week?

**Comment by asimshankar** [comment]
I'll reach out to you guys over email to figure out a time this week.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by asimshankar** [comment]
Closing this as per offline discussion with @cvinson830  - where the general idea was that this seems like it can be done an an independent TensorBoard plugin, without needing any changes to TensorFlow.

Feel free to reopen if I'm mistaken.

**Comment by ceevee830** [comment]
Thanks @asimshankar Yes we abstracted our data into a separate protobuf, thus decoupling from existing TF protobufs.
